src/Lecture2.hs

```diff
@@ -150,14 +170,219 @@ You're free to define any helper functions.
        treasure besides gold (if you already haven't done this).
 -}
 
--- some help in the beginning ;)
+newtype Gold = Gold Int
+newtype XP = XP Int
+newtype Health = Health Int deriving (Eq, Ord)
+newtype Attack = Attack Int deriving (Eq, Ord)
+newtype Endurance = Endurance Int deriving (Eq, Ord)
+
+instance Num Health where
+    (+) (Health a) (Health b) = Health (a + b)
+    (*) (Health a) (Health b) = Health (a * b)
+    abs (Health x) = Health (abs x)
+    signum (Health x) = Health (signum x)
+    (-) (Health a) (Health b) = Health (a - b)
+    fromInteger = Health . fromInteger
+
+instance Num Attack where
+    (+) (Attack a) (Attack b) = Attack (a + b)
+    (*) (Attack a) (Attack b) = Attack (a * b)
+    abs (Attack x) = Attack (abs x)
+    signum (Attack x) = Attack (signum x)
+    (-) (Attack a) (Attack b) = Attack (a - b)
+    fromInteger = Attack . fromInteger
+
+instance Num Endurance where
+    (+) (Endurance a) (Endurance b) = Endurance (a + b)
+    (*) (Endurance a) (Endurance b) = Endurance (a * b)
+    abs (Endurance x) = Endurance (abs x)
+    signum (Endurance x) = Endurance (signum x)
+    (-) (Endurance a) (Endurance b) = Endurance (a - b)
+    fromInteger = Endurance . fromInteger
+
+healthToInt :: Health -> Int
+healthToInt (Health h) = h
+
+attackToInt :: Attack -> Int
+attackToInt (Attack atk) = atk
+
+instance Show Gold where
+    show (Gold g) = "💰 " ++ show g
+
+instance Show XP where
+    show (XP xp) = "✨ " ++ show xp
+
+instance Show Health where
+    show (Health h) = "💖 " ++ show h
+
+instance Show Attack where
+    show (Attack atk) = "🔪 " ++ show atk
+
+instance Show Endurance where
+    show (Endurance e) = "✌ " ++ show e
+
+data Treasure = Treasure
+
+class ChestT a where
+    chestGold :: a -> Gold
+    chestTreasure :: a -> Maybe Treasure
+
+newtype ChestWithTreasureT = ChestWithTreasureT (Gold, Treasure)
```

Comment:
`newtype` over a pair usually doesn't have any benefits comparing to `data` with two fields. Especially if you can define `data` as a record